polygeist/test/polygeist-opt/sycl/licm-accessor-versioning.mlir

```diff
@@ -91,20 +92,126 @@ func.func private @test(%arg0: memref<?x!sycl_accessor_1_i32_rw_gb, 4>, %arg1: m
     affine.store %3, %alloca_2[0] : memref<1x!sycl_id_1_>
     %4 = sycl.accessor.subscript %arg1[%cast_3] {ArgumentTypes = [memref<?x!sycl_accessor_1_i32_r_gb, 4>, memref<?x!sycl_id_1_>], FunctionName = @"operator[]", TypeName = @accessor} : (memref<?x!sycl_accessor_1_i32_r_gb, 4>, memref<?x!sycl_id_1_>) -> memref<?xi32, 4>
     %5 = affine.load %4[0] : memref<?xi32, 4>
-    %6 = arith.extui %2 : i32 to i64
-    sycl.constructor @id(%memspacecast_6, %6) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl_id_1_, 4>, i64)
+    sycl.constructor @id(%memspacecast_6, %2) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl_id_1_, 4>, i64)
+    %6 = affine.load %alloca_0[0] : memref<1x!sycl_id_1_>
+    affine.store %6, %alloca[0] : memref<1x!sycl_id_1_>
+    %7 = sycl.accessor.subscript %arg0[%cast] {ArgumentTypes = [memref<?x!sycl_accessor_1_i32_rw_gb, 4>, memref<?x!sycl_id_1_>], FunctionName = @"operator[]", TypeName = @accessor} : (memref<?x!sycl_accessor_1_i32_rw_gb, 4>, memref<?x!sycl_id_1_>) -> memref<?xi32, 4>
+    %8 = affine.load %7[0] : memref<?xi32, 4>
+    %9 = arith.addi %8, %5 : i32
+    affine.store %9, %7[0] : memref<?xi32, 4>
+  }
+  return
+}
+
+// -----
+
+// This test requires -licm-sycl-accessor-pairs-limit >= 2.
+// Original loop:
+// for(size_t i = 0; i < 8; i++) {
+//   A[0] = 1;
+//   B[i] = 2;
+//   C[i] = 3;
+// }
+// Optimized loop:
+// if (0 < 8) {
+//   if ((&A[A.get_range()] <= &B[0] || &A[0] >= B[B.get_range()])
+//       && (&A[A.get_range()] <= &C[0] || &A[0] >= C[C.get_range()])) {
+//     A[0] = 1;
+//     for(size_t i = 0; i < 8; i++) {
+//       B[i] = 2;
+//       C[i] = 3;
+//     }
+//   } else {
+//     for(size_t i = 0; i < 8; i++) {
+//       A[0] = 1;
+//       B[i] = 2;
+//       C[i] = 3;
+//     }
+//   }
+// }
+
+!sycl_id_1_ = !sycl.id<[1], (!sycl.array<[1], (memref<1xi64>)>)>
+!sycl_range_1_ = !sycl.range<[1], (!sycl.array<[1], (memref<1xi64>)>)>
+!sycl_accessor_impl_device_1_ = !sycl.accessor_impl_device<[1], (!sycl_id_1_, !sycl_range_1_, !sycl_range_1_)>
+!sycl_accessor_1_i32_w_gb = !sycl.accessor<[1, i32, write, global_buffer], (!sycl_accessor_impl_device_1_, !llvm.struct<(memref<?xi32, 1>)>)>
+
+// COM: Store to %arg0 accessor cannot be hoisted.
+// 1PAIR-LABEL: test
+// 1PAIR-SAME:  ([[ARG0:%.*]]: memref<?x[[ACC_W:!sycl_accessor_1_i32_w_gb]], 4>, [[ARG1:%.*]]: memref<?x[[ACC_W]], 4>, [[ARG1:%.*]]: memref<?x[[ACC_W]], 4>) 
+// 1PAIR: [[C1_i32:%.*]] = arith.constant 1 : i32
+// 1PAIR: [[ARG0_ACC:%.*]] = sycl.accessor.subscript %arg0[{{.*}}] {ArgumentTypes = [memref<?x!sycl_accessor_1_i32_w_gb, 4>, memref<?x!sycl_id_1_>], FunctionName = @"operator[]", TypeName = @accessor} : (memref<?x!sycl_accessor_1_i32_w_gb, 4>, memref<?x!sycl_id_1_>) -> memref<?xi32, 4>
+// 1PAIR: affine.store [[C1_i32]], [[ARG0_ACC]][0] : memref<?xi32, 4>
+// 1PAIR-NOT: scf.for
```

Comment:
This is confusing because even if we aren't versioning the loop at all the original `scf.for` loop should still be present.
The test should check that the original scf.for is generated and (then) check that after that there aren't any more `scf.for` loops.